### PR TITLE
Implement transaction_data binding verification for mso_mdoc credentials

### LINKED ISSUE #124: Implement transaction_data binding verification for mso_mdoc credentials
Follow up to:
- #88 

Self-explanatory. Covered for SD-JWT but not mDoc.

ETA: 1 day.

### PR DESCRIPTION
**Pull request by Awambeng**
This PR adds `transaction_data_hashes` validation for `mso_mdoc` credentials, analogous to the existing SD-JWT KB-JWT validation.

The wallet includes the hashes in `DeviceSigned.nameSpaces`, where they are covered by the device signature. During presentation verification, the verifier extracts these hashes and validates them against the transaction data supplied in the
authorization request.

### Changes

- **VerifierConfig**: Removed the guard preventing `transaction_data` from being used with `mso_mdoc` as the primary credential.
- **MdocVerificationContext**: Added `getDeviceNameSpaces()` to expose `DeviceSigned.nameSpaces` during verification.
- **MdocCredentialVerifier**: Added `validateTransactionData()`, which reads `transaction_data_hashes` from the device namespaces and delegates validation to `TransactionDataValidator`.
- **MdocBaseTest**: Added builder support for custom `DeviceNameSpaces` in test mDocs.
- **Tests**: Updated the profile configuration test and added four transaction data verification tests for mDoc credentials.

Closes #124

### CODE REVIEW

**Comment by Awambeng** [comment]
> ## Code Review Summary
> ### Critical Issue
> **Algorithm Confusion Attack** (`MdocCredentialVerifier.java:178-181`): The hash algorithm is extracted from the client-supplied mDoc and passed to `TransactionDataValidator`. An attacker could use a weaker algorithm (e.g., MD5) to circumvent transaction data binding security. The verifier should dictate the expected algorithm, not trust the client's `TRANSACTION_DATA_HASHES_ALG_CLAIM`.
> 
The hash algorithm is validated against both the verifier's request-specified `transaction_data_hashes_alg` (which dictates the allowed set) and `TransactionDataSupport.validateHashAlgorithm()`  which only accepts  sha-256  (hardcoded). The wallet's claimed algorithm must match the verifier's allowed algorithms. No attack vector exists with the current supported algorithms.

> ### Medium Issues
> 1. **Silent Hash Filtering** (`extractHashesArray()`): Non-`CBORString` items are dropped without warning or error. Malformed data silently reduces validation scope.
>
Fixed. `VerificationException` is thrown when encountering non-string entries in the hashes array, preventing silent hash filtering.

> 2. **Namespace Loop Overwrites** (`validateTransactionData()`): If `transaction_data_hashes` appears in multiple namespaces, only the last one's hashes are used. This could allow an attacker to hide hashes in multiple namespaces.
> 
Fixed. We now reject transaction data hashes appearing in multiple namespaces with a descriptive error.

> ### Missing Test Coverage
> * Algorithm mismatch scenarios
> * Malformed CBOR hash structures
> * End-to-end `verifyCredential()` flow with both revocation and transaction data
> 
Fixed. Added test cases for:
- Malformed CBOR hashes
- Duplicate namespaces
- End-to-end test for revocation and transaction data

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java [hunk 1]
```diff
@@ -95,9 +112,110 @@ public JsonNode verifyCredential(
...
+        }
+
+        ObjectNode hashesPayload = extractTransactionDataHashes(verificationContext);
+        if (hashesPayload == null) {
+            throw new VerificationException(
+                    "Device namespaces must contain transaction_data_hashes when transaction_data is requested");
+        }
+
+        try {
+            TransactionDataValidator.validate(transactionDataWire, hashesPayload);
```
Comment (forkimenjeckayang) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java:148]: > This proves that the hashes are covered by DeviceAuth, but it does not prove that the issuer authorized the device key to sign this transaction-data element. Final 1.0 Appendix B.2.1 defines mDoc transaction-data binding through a transaction-type-specific `(NameSpace, DataElementIdentifier, DataElementValue)` and says the issuer includes the relevant element in `DeviceKeyInfo.keyAuthorizations`. The verifier currently scans for the SD-JWT name `transaction_data_hashes` in any namespace and never inspects `keyAuthorizations` or maps the requested `type` to an mDoc transaction-data profile.
>
> The positive fixture demonstrates the gap: `MdocBaseTest` authorizes only the issuer-claim namespace `com.example.namespace1`, while `buildMdocWithTransactionData()` places the hashes under `com.example.doctype`; the value is accepted despite that namespace not being issuer-authorized.
>
Please model the supported mDoc transaction-data profile(s), mapping each supported request `type` to its exact namespace, element identifier, value structure, and processing/hash rules. Before accepting the binding, require that exact element or its namespace is present in the issuer-signed MSO `KeyAuthorizations`; reject unconfigured types or unauthorized elements. Add full-verifier tests for authorized, unauthorized, missing, and mismatched elements. Reusing the shared hash primitive is fine once the mDoc profile has selected and authorized the value.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java [hunk 2]
```diff
@@ -95,9 +112,110 @@ public JsonNode verifyCredential(
...
+            if (hashesPair != null) {
+                if (hashes != null) {
+                    throw new VerificationException("transaction_data_hashes must not appear in multiple namespaces");
+                }
+                hashes = extractHashesArray(hashesPair.getValue());
+            }
+            var algPair = itemsMap.findByKey(TransactionDataValidator.TRANSACTION_DATA_HASHES_ALG_CLAIM);
+            if (algPair != null && algPair.getValue() instanceof CBORString algStr) {
+                hashesAlg = algStr.getValue();
+            }
```
Comment (forkimenjeckayang) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java:196]: > This loop loses the namespace association required by Appendix B.2.1. It can take `transaction_data_hashes` from namespace A and `transaction_data_hashes_alg` from namespace B, then synthesize one payload that never existed as a single mDoc data-element tuple. Multiple algorithm entries are also silently resolved by taking the last one.
>
 Please resolve one exact profile namespace first and read all related elements only from that namespace. Reject duplicate, cross-namespace, wrong-type, or unexpected algorithm elements. Add regressions where the hash and algorithm are split across namespaces and where multiple namespaces supply an algorithm.

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocRevocationStatusTest.java [hunk 3]
```diff
@@ -141,6 +150,77 @@ public void shouldFail_WhenRevocationEnforcedAndStatusInvalid() throws Exception
...
+                .thenReturn(JsonSerialization.writeValueAsString(List.of(wire)));
+
+        var context = mock(AuthenticationFlowContext.class);
+        when(context.getAuthenticatorConfig()).thenReturn(authConfig);
+        when(context.getAuthenticationSession()).thenReturn(authSession);
+
+        var credential = new CredentialRequirement()
+                .setId("test")
+                .setCredentialTypes(List.of(DOC_TYPE))
+                .setTrust(List.of(new TrustPolicy().setType(TrustPolicy.X5C).setAnchors(List.of(getIssuerCertRef1()))));
```
Comment (forkimenjeckayang) [inline src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocRevocationStatusTest.java:190]: > This credential keeps the default `SUPPORTING` role, but production calls `validateTransactionData()` only when `credential.isPrimary()` is true. Consequently, `shouldVerifyCredential_WithBothRevocationAndTransactionData` verifies revocation only; removing or breaking the new transaction-data call would not fail this test. The other new tests invoke the package-private helper directly without first verifying issuer signature, DeviceAuth, or the MSO.
>
 Please set this requirement to `CredentialRole.PRIMARY` and assert that a mismatched transaction hash is rejected through `verifyCredential()`. The positive full-flow fixture should also carry a KeyAuthorization for the selected transaction-data element, with a negative fixture proving that an otherwise valid DeviceSignature is rejected when that authorization is absent.

**Review by forkimenjeckayang — COMMENTED** [review]
a few comments

File: src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocRevocationStatusTest.java (hunk 3 above)
Comment (Awambeng) [reply src/test/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocRevocationStatusTest.java:190]: Thanks. Changes applied as suggested.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java (hunk 2 above)
Comment (Awambeng) [reply src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java:196]: Refactored to a two-pass approach: first resolves the single namespace containing  transaction_data_hashes , then reads  transaction_data_hashes_alg  only from that namespace. Rejects alg in any other namespace, rejects non-string alg values, and handles both bare CBOR string and single-element array encodings (authlete  DeviceSignedItemsEntry  wrapping) via  extractAlgString() . Regression test added for split alg/hashes namespaces.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java (hunk 1 above)
Comment (Awambeng) [reply src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java:148]: Added `verifyNamespaceIsAuthorized()` checks that the namespace containing `transaction_data_hashes` is present in the MSO's `deviceKeyInfo.keyAuthorizations.nameSpaces` before accepting the binding. This addresses the issuer authorization gap with positive (authorized namespace) and negative (unauthorized namespace) verifier tests. Broader transaction-data profile mapping can be introduced separately if needed.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java [hunk 4]
```diff
@@ -95,9 +113,168 @@ public JsonNode verifyCredential(
...
+        for (JsonNode ns : authorizedNs) {
+            if (ns.asText().equals(namespace)) {
+                authorized = true;
+                break;
+            }
+        }
+        if (!authorized) {
+            throw new VerificationException(
+                    "Namespace '" + namespace + "' is not authorized for transaction_data_hashes in the MSO");
+        }
```
Comment (forkimenjeckayang) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java]: > This now checks MSO `KeyAuthorizations`, but only the whole-namespace form. ISO mdoc also permits `keyAuthorizations.dataElements`, where the device key is authorized for named elements without being authorized for the entire namespace. A credential whose MSO authorizes `transaction_data_hashes` exactly through `dataElements` is therefore rejected as missing authorized `nameSpaces`.

Please accept an element when either its namespace is listed in `nameSpaces` or the element identifier is listed under `dataElements[namespace]`. Apply that check to every DeviceSigned element consumed here: `transaction_data_hashes`, and `transaction_data_hashes_alg` when it is present. Add a positive `dataElements`-only test and negative tests where the hash or algorithm element is not authorized.

**Review by forkimenjeckayang — COMMENTED** [review]
one last comment

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java (hunk 4 above)
Comment (Awambeng) [reply src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java]: fixed.

**Review by forkimenjeckayang — DISMISSED** [review]
Thanks for addressing my concerns

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/mdoc/MdocVerificationContext.java [hunk 5]
```diff
@@ -59,6 +59,16 @@ public MdocVerificationContext(String mdoc) throws VerificationException {
...
 
+    /**
+     * Returns the DeviceSigned namespaces from the parsed mDoc for use during
+     * presentation verification, including inspection of elements such as
+     * {@code transaction_data_hashes}.
+     */
+    public CBORTaggedItem getDeviceNameSpaces() throws VerificationException {
+        CBORPairList document = extractDocument(mdoc);
+        return extractDeviceNamespaces(document);
+    }
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/mdoc/MdocVerificationContext.java]: Could you relocate this getter next to `getVerifiedMsoPayload`? It helps readability to keep similar methods close. Also, this parses the mdoc tree once again. Could you check if it's possible to use the same pattern with the MSO payload?

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java [hunk 6]
```diff
@@ -96,9 +109,198 @@ public JsonNode verifyCredential(
             }
         }
 
+        if (credential.isPrimary()) {
+            validateTransactionData(authorizationContext, verificationContext);
+        }
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java]: Because both verifiers do this, what about adding `validateTransactionData` to the contract of a CredentialVerifier and call this directly in OID4VPAuthenticator? So we're sure future formats do not omit to make this call.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java [hunk 7]
```diff
@@ -96,9 +109,198 @@ public JsonNode verifyCredential(
...
+            var algPair = itemsMap.findByKey(TransactionDataValidator.TRANSACTION_DATA_HASHES_ALG_CLAIM);
+            if (algPair == null) {
+                continue;
+            }
+            if (!nsName.equals(hashesNamespace)) {
+                throw new VerificationException(
+                        "transaction_data_hashes_alg must not appear outside the namespace containing transaction_data_hashes");
+            }
+            hashesAlg = extractAlgString(algPair.getValue());
+        }
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java]: Could you optimize these passes into just one? For each potential namespace, you collect both values while checking that the current namespace is equal to hashesNamespace if not null. Or a similar logic.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java [hunk 8]
```diff
@@ -96,9 +109,198 @@ public JsonNode verifyCredential(
...
+        if (dki == null) {
+            throw new VerificationException("MSO is missing deviceKeyInfo");
+        }
+        JsonNode keyAuth = dki.get("keyAuthorizations");
+        if (keyAuth == null) {
+            throw new VerificationException("MSO deviceKeyInfo is missing keyAuthorizations");
+        }
+        // Check namespace-level authorization first
+        JsonNode nameSpaces = keyAuth.get("nameSpaces");
+        if (nameSpaces != null && nameSpaces.isArray()) {
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java:258]: Use constants from MdocConstants. If not available, add them.

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java [hunk 9]
```diff
@@ -96,9 +109,198 @@ public JsonNode verifyCredential(
...
+                    }
+                    if (!found) {
+                        allFound = false;
+                        break;
+                    }
+                }
+                if (allFound) {
+                    return;
+                }
+            }
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/mdoc/MdocCredentialVerifier.java:276]: These nested loops are not very friendly nor efficient. Could you explore leveraging the Java Stream API to achieve the same?

```java
if (elements != null && elements.isArray()) {
    Set<String> existingIds = StreamSupport.stream(elements.spliterator(), false)
            .map(JsonNode::asText)
            .collect(Collectors.toSet());

    if (existingIds.containsAll(elementIds)) {
        return;
    }
}
```

**Review by IngridPuppet — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/OID4VPAuthenticator.java [hunk 10]
```diff
@@ -86,6 +86,10 @@ public void authenticate(AuthenticationFlowContext context) {
         try {
             primaryClaims = primaryVerifier.verifyCredential(
                     context, authContext, authRequirements, primaryCredential, primaryToken);
+            // SD-JWT validates transaction_data via the interface method (above).
+            // mso_mdoc validates internally inside verifyCredential() because it
+            // depends on the post-verification MdocVerificationContext state.
+            primaryVerifier.validateTransactionData(authContext, primaryToken);
```
Comment (IngridPuppet) [inline src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/OID4VPAuthenticator.java]: This behavior is strange. All implementations must behave the same. Can't MdocCredentialVerifier just keep the post verification context as a property in the verifier? Also there should be no default implementation of `validateTransactionData` in the interface, so a no-op impl must be explicit not implicit.

**Review by IngridPuppet — COMMENTED** [review]
Just one last architectural comment...

File: src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/OID4VPAuthenticator.java (hunk 10 above)
Comment (Awambeng) [reply src/main/java/io/github/adorsysgis/keycloak/protocol/oid4vc/oid4vp/authenticator/OID4VPAuthenticator.java]: Addressed.

- Removed the default validateTransactionData implementation. It is now required for all implementations.
- MdocCredentialVerifier now stores the MdocVerificationContext during verifyCredential() and uses it for transaction validation. Removed the credential.isPrimary() check.

**Review by IngridPuppet — APPROVED** [review]
Thank you very much for the update.
